Cron/ProductsTopic.php

```diff
@@ -0,0 +1,123 @@
+     * Klaviyo Logger
+     * @var Logger
+     */
+    protected $_klaviyoLogger;
+
+    /**
+     * Magento product category helper
+     * @var CategoryFactory $categoryFactory
+     */
+    protected $_categoryFactory;
+
+    /**
+     * Klaviyo Products Resource Model
+     * @var Products
+     */
+    protected $_klProduct;
+
+    /**
+     * Klaviyo Products Collection
+     * @var CollectionFactory
+     */
+    protected $_klProductCollectionFactory;
+
+    /**
+     * Klaviyo Syncs Model
+     * @var SyncsFactory
+     */
+    protected $_klSyncFactory;
+
+    /**
+     * @param Logger $klaviyoLogger
+     * @param Products $klProduct
+     * @param SyncsFactory $klSyncFactory
+     * @param CollectionFactory $klProductCollectionFactory
+     */
+    public function __construct(
+        Logger $klaviyoLogger,
+        Products $klProduct,
+        CategoryFactory $categoryFactory,
+        SyncsFactory $klSyncFactory,
+        CollectionFactory $klProductCollectionFactory
+    )
+    {
+        $this->_klaviyoLogger = $klaviyoLogger;
+        $this->_klProduct = $klProduct;
+        $this->_categoryFactory = $categoryFactory;
+        $this->_klSyncFactory = $klSyncFactory;
+        $this->_klProductCollectionFactory = $klProductCollectionFactory;
+    }
+
+    public function queueKlProductsForSync()
+    {
+        $klProductsCollection = $this->_klProductCollectionFactory->create();
+        $klProductsToSync = $klProductsCollection->getRowsForSync('NEW')
+            ->addFieldToSelect(['id','payload','status','topic', 'klaviyo_id'])
+            ->getData();
+
+        if (empty($klProductsToSync)) {return;}
+
+        $idsToUpdate = [];
+
+        foreach ($klProductsToSync as $klProductToSync)
+        {
+            $klProductToSync['payload'] = json_encode($this->addCategoryNames($klProductToSync['payload']));
+            $klSync = $this->_klSyncFactory->create();
+            $klSync->setData([
+                'payload'=> $klProductToSync['payload'],
+                'topic'=> $klProductToSync['topic'],
+                'klaviyo_id'=>$klProductToSync['klaviyo_id'],
+                'status'=> 'NEW'
+            ]);
+            try {
+                $klSync->save();
+                array_push($idsToUpdate, $klProductToSync['id']);
+            } catch (\Exception $e) {
+                $this->_klaviyoLogger->log(sprintf('Unable to move row: %s', $e->getMessage()));
+            }
+        }
+
+        $klProductsCollection->updateRowStatus($idsToUpdate, 'MOVED');
+    }
+
+    public function clean()
+    {
+        $klProductsCollection = $this->_klProductCollectionFactory->create();
+        $idsToDelete = $klProductsCollection->getIdsToDelete('MOVED');
+
+        $klProductsCollection->deleteRows($idsToDelete);
+    }
+
+    /**
+     * Helper function to associate category names with their ids
+     * @param string $payload
+     * @return array
+     */
+    public function addCategoryNames(string $payload): array
```

Comment:
Over here, we will be iterating over each categoryId for all payloads passed, so we will be running the for loop 500 times, if categoryIds, exist and possibly loading duplicate categroyModels since it is likely more than one product will have a category repeated. Can we make it such that we create a class level attribute to have Ids and categoryNames associated so don't duplicate the effort? 

https://github.com/klaviyo/magento2-klaviyo/pull/142#discussion_r742024644 > as advised here? 

Or do you think it is worth having a method in the helper class which can be used by both the Cron?